src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -265,12 +280,223 @@ private RouteFW resolveTarget(
+    }
+
+    private void doSignal(
+        MessageConsumer receiver,
+        long routeId,
+        long streamId,
+        long traceId)
+    {
+        final SignalFW signal = signalRW.wrap(writeBuffer, 0, writeBuffer.capacity()).routeId(routeId)
+                                        .streamId(streamId)
+                                        .traceId(traceId)
+                                        .build();
+
+        receiver.accept(signal.typeId(), signal.buffer(), signal.offset(), signal.sizeof());
+    }
+
+    private int decodeConnectPacket(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final MqttPacketType packetType,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        int progress = offset;
+        switch (packetType)
+        {
+        case CONNECT:
+            MqttConnectFW mqttConnect = mqttConnectRO.wrap(buffer, offset, limit);
+            server.onMqttConnect(authorization, mqttConnect);
+            progress = mqttConnect.limit();
+            break;
+        default:
+            server.doNetworkEnd(traceId);
+            break;
+        }
+        return progress;
+    }
+
+    private int decodeSession(
+        MqttServer server,
+        final long traceId,
+        final long authorization,
+        final MqttPacketType packetType,
+        final DirectBuffer buffer,
+        final int offset,
+        final int limit)
+    {
+        int progress = 0;
+
+        switch (packetType)
+        {
+        case PINGREQ:
+            final MqttPingReqFW ping = mqttPingReqRO.wrap(buffer, offset, limit);
+            server.onMqttPingReq(authorization, ping);
+            progress = ping.limit();
+            break;
+        case SUBSCRIBE:
+            final MqttSubscribeFW subscribe = mqttSubscribeRO.wrap(buffer, offset, limit);
+            server.onMqttSubscribe(authorization, subscribe);
+            progress = subscribe.limit();
+            break;
+        case UNSUBSCRIBE:
+            final MqttUnsubscribeFW unsubscribe = mqttUnsubscribeRO.wrap(buffer, offset, limit);
+            server.onMqttUnsubscribe(authorization, unsubscribe);
+            progress = unsubscribe.limit();
+            break;
+        case DISCONNECT:
+            final MqttDisconnectFW disconnect = mqttDisconnectRO.wrap(buffer, offset, limit);
+            server.onMqttDisconnect(disconnect);
+            server.doNetworkEnd(traceId);
+            progress = disconnect.limit();
+            break;
+        case PUBLISH:
+            final MqttPublishFW publish = mqttPublishRO.wrap(buffer, offset, limit);
+            server.onMqttPublish(authorization, publish);
+            progress = publish.limit();
+            break;
+        default:
+            server.doNetworkReset(traceId);
+            break;
+        }
+
+        return progress;
+    }
```

Comment:
Change to `decodePacketType`, then based on observed `packetType`, change `decoder` to packet specific, such as `decodePing` or `decodeSubscribe`.